graylog2-server/src/main/java/org/graylog/events/notifications/types/HTTPEventNotification.java

```diff
@@ -63,18 +65,22 @@ public interface Factory extends EventNotification.Factory {
     private final UrlWhitelistService whitelistService;
     private final UrlWhitelistNotificationService urlWhitelistNotificationService;
     private final EncryptedValueService encryptedValueService;
+    private EventsConfigurationProvider configurationProvider;
 
     @Inject
     public HTTPEventNotification(EventNotificationService notificationCallbackService, ObjectMapper objectMapper,
-                                 final OkHttpClient httpClient, UrlWhitelistService whitelistService,
+                                 final OkHttpClientProvider httpClient, UrlWhitelistService whitelistService,
                                  UrlWhitelistNotificationService urlWhitelistNotificationService,
-                                 EncryptedValueService encryptedValueService) {
+                                 EncryptedValueService encryptedValueService,
+                                 EventsConfigurationProvider configurationProvider
+                                 ) {
         this.notificationCallbackService = notificationCallbackService;
         this.objectMapper = objectMapper;
-        this.httpClient = httpClient;
+        this.httpClient = httpClient.getWithTcpKeepAlive(socket -> configurationProvider.get().notificationsKeepAliveProbe());
```

Comment:
using the provider directly will always create a new http client instance. This renders http keep alive useless,
because it cannot reuse the pool of the previous client.

We could cache this manually. (And rebuild it when the keep alive setting changes)
E.g.: 
```
diff --git graylog2-server/src/main/java/org/graylog2/shared/bindings/providers/OkHttpClientProvider.java graylog2-server/src/main/java/org/graylog2/shared/bindings/providers/OkHttpClientProvider.java
index a032b7ee3f..a719eae193 100644
--- graylog2-server/src/main/java/org/graylog2/shared/bindings/providers/OkHttpClientProvider.java
+++ graylog2-server/src/main/java/org/graylog2/shared/bindings/providers/OkHttpClientProvider.java
@@ -74,6 +74,8 @@ public class OkHttpClientProvider implements Provider<OkHttpClient> {
     protected final URI httpProxyUri;
     protected final ProxyHostsPattern nonProxyHostsPattern;
 
+    private OkHttpClient cachedHttpKeepAliveClient;
+
     @Inject
     public OkHttpClientProvider(@Named("http_connect_timeout") Duration connectTimeout,
                                 @Named("http_read_timeout") Duration readTimeout,
@@ -141,9 +143,12 @@ public class OkHttpClientProvider implements Provider<OkHttpClient> {
     }
 
     public OkHttpClient getWithTcpKeepAlive(Predicate<Socket> keepAlivePredicate) {
-        return get().newBuilder()
-                .socketFactory(new TcpKeepAliveSocketFactory(SocketFactory.getDefault(), keepAlivePredicate))
-                .build();
+        if (cachedHttpKeepAliveClient == null) {
+            cachedHttpKeepAliveClient = get().newBuilder()
+                    .socketFactory(new TcpKeepAliveSocketFactory(SocketFactory.getDefault(), keepAlivePredicate))
+                    .build();
+        }
+        return cachedHttpKeepAliveClient;
     }
 
     public static class ProxyAuthenticator implements Authenticator {

```

Also I'm not a big fan of the Predicate. Do we need this? All I see is more complexity

Reply:
I wrongly assumed that the class is a singleton. You are right, now the implementation is preventing the reuse of the client. Let me refactor that once again.